File: examples_py/tests/generators/test_gcopter.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Unit test for :class:`examples_py.generators.GcopterGenerator`."""

import unittest

import numpy as np

from examples_py.framework import ReferenceField, has_field
from examples_py.generators import GcopterGenerator
from _helpers import (
    hover_state_at,
    load_test_sim_config,
    repo_path,
)


class GcopterGeneratorTest(unittest.TestCase):
    def setUp(self) -> None:
        self.sim_cfg = load_test_sim_config()
        self.state = hover_state_at([0.0, 0.0, 10.0])
        cfg = GcopterGenerator.load_config_from_yaml(
            str(repo_path('configs/generators/config_gcopter.yaml')))
        self.gen = GcopterGenerator(cfg)
        self.gen.initialize(self.state, self.sim_cfg)

    def test_produces_finite_references_over_50_steps(self) -> None:
        self.gen.on_waypoint_changed([5.0, 0.0, 10.0], self.state, 0.0)
        for k in range(50):
            t = 0.01 * k
            self.gen.update(t, self.state)
            s = self.gen.evaluate(t)
            self.assertTrue(np.all(np.isfinite(s.position)), f'step {k}')

    def test_advertises_trajectory_fields(self) -> None:
        mask = self.gen.provided_reference_fields()
        self.assertTrue(has_field(mask, ReferenceField.POSITION))


if __name__ == '__main__':  # pragma: no cover
    unittest.main()

File: examples_cpp/tests/generators/test_gcopter.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

#include <gtest/gtest.h>

#include <memory>

#include "framework/trajectory_generator_base.hpp"
#include "generators/gcopter_generator.hpp"
#include "test_helpers.hpp"

using mpc_examples::adapters::GcopterGenerator;
using mpc_examples::testing::hoverStateAt;
using mpc_examples::testing::loadTestSimConfig;
using mpc_examples::testing::repoPath;

class GcopterGeneratorTest : public ::testing::Test {
protected:
  mpc_examples::ExampleConfig sim_cfg_ = loadTestSimConfig();
  mav_model::State            state_   = hoverStateAt({0.0, 0.0, 10.0});
  std::unique_ptr<GcopterGenerator> gen_;

  void SetUp() override {
    auto cfg = GcopterGenerator::loadConfigFromYaml(
        repoPath("configs/generators/config_gcopter.yaml"));
    gen_ = std::make_unique<GcopterGenerator>(cfg);
    gen_->initialize(state_, sim_cfg_);
  }
};

TEST_F(GcopterGeneratorTest, ProducesFiniteReferencesOver50Steps) {
  const Eigen::Vector3d target(5.0, 0.0, 10.0);
  gen_->onWaypointChanged(target, state_, 0.0);

  for (int k = 0; k < 50; ++k) {
    const double t = 0.01 * k;
    gen_->update(t, state_);
    const auto s = gen_->evaluate(t);
    ASSERT_TRUE(s.position.allFinite()) << "step " << k;
    ASSERT_TRUE(s.velocity.allFinite()) << "step " << k;
  }
}

TEST_F(GcopterGeneratorTest, AdvertisesTrajectoryFields) {
  const auto mask = gen_->providedReferenceFields();
  using mpc_examples::framework::ReferenceField;
  using mpc_examples::framework::hasField;
  EXPECT_TRUE(hasField(mask, ReferenceField::kPosition));
  EXPECT_TRUE(hasField(mask, ReferenceField::kVelocity));
}
